Multiple registrations

pbetkier opened the pull request:
This PR enabled registering multiple service instances for a given marathon app, e.g. with different ports, names or tags. These additional registration details can be provided in `portDefinitions` part of the marathon app configuration. Previous configuration method, via app-level `labels`, is still supported and that has not changed.

I'm going to commit changes in README on Monday ;)

janisz commented on apps/app.go:
```diff
@@ -49,8 +58,20 @@ func (app *App) IsConsulApp() bool {
-func (app *App) ConsulName() string {
-	if value, ok := app.Labels[MARATHON_CONSUL_LABEL]; ok && !isSpecialConsulNameValue(value) {
+func (app *App) HasSameConsulNamesAs(other *App) bool {
+	return reflect.DeepEqual(rawConsulNames(app), rawConsulNames(other))
```
I prefer to not use `reflect`. How do you guarantee `rawConsulNames` for two different structs will return names in same order.

pbetkier replied:
Well, the implementation is deterministic so if configurations are the same then names will also be.

janisz replied:
Got it! I'll prefer to replace reflect with simple loop and do not rely on `reflect` package but I don't insist.

janisz commented on apps/app.go:
```diff
@@ -20,6 +24,10 @@ type HealthCheck struct {
 }
 
+type PortDefinition struct {
+	Labels map[string]string `json:"labels"`
```
No `port` field?

janisz commented on apps/app.go:
```diff
@@ -73,3 +94,77 @@ func ParseApp(jsonBlob []byte) (*App, error) {
+	for i, d := range app.PortDefinitions {
+		if _, ok := d.Labels[MARATHON_CONSUL_LABEL]; ok {
+			definitions = append(definitions, IndexedPortDefinition{
+				PortIndex: i,
```
Why you assume port definitions are sorted by port index?

pbetkier replied:
In `portDefinitions` you can either provide a port explicitly or provide zero in which case the port will be randomly picked. I've added support for the latter only. I guess you're right that we should support both ways.

wendigo replied:
Guys, it's saturday! Get rest :)

janisz replied:
Open source never sleeps.

wendigo replied:
✊

janisz commented on apps/app_test.go:
```diff
@@ -119,3 +119,215 @@ func TestAppId_String(t *testing.T) {
 	assert.Equal(t, "appId", AppId("appId").String())
 }
+
+func TestRegistrationIntent_NameWithSeparator(t *testing.T) {
```
Think about using [TableDrivenTests](https://github.com/golang/go/wiki/TableDrivenTests) or

janisz replied:
Think about using [TableDrivenTests](https://github.com/golang/go/wiki/TableDrivenTests)

janisz commented on apps/app.go:
```diff
@@ -49,8 +58,20 @@ func (app *App) IsConsulApp() bool {
 
-func (app *App) ConsulName() string {
-	if value, ok := app.Labels[MARATHON_CONSUL_LABEL]; ok && !isSpecialConsulNameValue(value) {
+func (app *App) HasSameConsulNamesAs(other *App) bool {
```
Missing unit test for this function.

pbetkier replied:
Added missing unit test.

medzin commented on consul/consul_test.go:
```diff
@@ -385,6 +386,66 @@ func TestRegisterServices_CustomServiceName(t *testing.T) {
+	// then
+	assert.Len(t, services, 2)
+	assert.Equal(t, "first-name", services[0].Name)
+	assert.Equal(t, []string{"marathon", "common-tag", "first-tag", "marathon-task:serviceA.0"}, services[0].Tags)
```
Why not just use [assert.Contains][https://godoc.org/github.com/stretchr/testify/assert#Contains] here?

pbetkier replied:
I would have to call assert.Contains 4 times and check the lengths to make sure it contains these tags and not other...

janisz replied:
Add dedicated method to compare slices so it will be clear

pbetkier replied:
How would it be different from checking if two slices are equal the way I did?

janisz replied:
No difference but it will be clear why sorting is required.

pbetkier replied:
Oh, we're speaking of 2 different things. Sorting is not for this line with tags, it's for picking the right services for assertions. I'll think how to rework it to make it more clear.

janisz replied:
How about comparing result of `consul.GetAllServices()` with expected list of services?

medzin replied:
Maybe GetAllServices should return sorted list, or it doesn't matter in not test code?

pbetkier commented:
I've added support for explicit ports, reworked checking if two apps have same names, added table-driven tests. What do you think now?

I'm working on the README at the moment, going to push that soon.

janisz commented on README.md:
```diff
@@ -215,14 +217,76 @@ Endpoint  | Description
 `/events` | event sink - returns `OK` if all keys are set in an event, error message otherwise
 
-### Known limitations
+## Advanced usage
```
How about naming this section "Multiple ports support" since it describes how to do it.

dankraw replied:
Or maybe we should add a sub section called 
```markdown
### Multiple ports support
```

dankraw commented on README.md:
````diff
@@ -215,14 +217,76 @@ Endpoint  | Description
+```
+  "id": "my-new-app",
+  "labels": {
+    "consul": "true",
````
Isn't `true` a *deprecated* value for this label?
It should be left blank if consul service name is going to be equal to marathon app name. 
Or we should remove the deprecation note if we can't get rid of this feature. :)

pbetkier replied:
Oh, ok. I can change it to empty. @janisz what do you think?

janisz replied:
I opt for using empty name. I think it's time to cleanup `consul: true` logic but not in this PR.

dankraw commented on README.md:
```diff
@@ -215,14 +217,76 @@ Endpoint  | Description
+      "protocol": "tcp",
+      "labels": {
+        "consul": "my-app-other-name",
+        "specific-tag": "tag
```
closing `"` missing

pbetkier replied:
damn, you got me :) ok, I confess, I wrote these examples by hand :(

dankraw replied:
still fine :)

dankraw commented on apps/app_test.go:
```diff
@@ -119,3 +119,375 @@ func TestAppId_String(t *testing.T) {
+	}
+
+	// expect
+	assert.True(t, app.HasSameConsulNamesAs(other))
```
the opposite way assertion `other.HasSameConsulNamesAs(app))` would be welcome :)

dankraw commented on consul/consul.go:
```diff
@@ -257,42 +269,28 @@ func (c *Consul) deregister(toDeregister *service.Service) error {
-}
-
-func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) (*consulapi.AgentServiceRegistration, error) {
+func (c *Consul) marathonTaskToConsulService(task *apps.Task, app *apps.App) ([]*consulapi.AgentServiceRegistration, error) {
```
renaming it to `marathonTaskToConsulServices` would be nice :)

dankraw commented on web/event_handler.go:
```diff
@@ -218,36 +218,13 @@ func (fh *eventHandler) handleDeploymentStepSuccess(body []byte) error {
 
-	for _, service := range services {
-		taskId, err := service.TaskId()
+	for _, task := range app.Tasks {
```
Are we sure the `app.Tasks` will be available? Like in this case from `event_handler.go`:

```go
	for _, app := range deploymentEvent.StoppedConsulApps() {
		for _, err = range fh.deregisterAllAppServices(app) {
			errors = append(errors, err)
		}
	}
```

pbetkier replied:
Oh I see, so that's why this logic was so complicated? Test missing! :)

dankraw replied:
`TestWebHandler_HandleDeploymentInfoWithStopApplicationAction` looks like a test that should fail here, but it might be wrong (generated json request body).

pbetkier replied:
Ok I fixed the test and the implementation, check this out @dankraw @janisz

janisz commented:
LGTM :+1:
